### PR TITLE
VPN-1839: Dinamically change status bar text color on mobile

### PR DESCRIPTION
**Pull request by brizental**
## Description

The default text color is black for this text. Contrary to what I
thought, iOS does not automatically verify background color and
change the text accordingly. Status bar text color needs to explicitly be
changed when necessary.

This is a draft because I still need to:

- [x] Stop using deprecated APIs on the iOS implementation
- [ ] ~Write tests~
   - Skipped for this functionality due to the lack of structue for adding platform specific tests at the moment.

I would like to get some early feedback on the approach I took here,
especially on how I handled this on the QML side, since this is my first PR on the project.
I thought it would be best to double check.

https://user-images.githubusercontent.com/25176023/164515154-2f296ad5-95b3-4228-b4b5-c0902815f451.mov

## Reference

https://mozilla-hub.atlassian.net/browse/VPN-1839

## Checklist
    
- [x] My code follows the style guidelines for this project
- [x] I have not added any packages that contain high risk or unknown licenses (GPL,  LGPL, MPL, etc. consult with DevOps if in question)
- [x] I have performed a self review of my own code
- [x] I have commented my code PARTICULARLY in hard to understand areas
- [ ] I have added thorough tests where needed

### CODE REVIEW

File: src/platforms/ios/iosutils.mm
Comment (brizental) [inline src/platforms/ios/iosutils.mm:122]: This is the deprecated APIs I mentioned on the PR description. They were deprecated on iOs 13.

File: src/ui/views/ViewMobileOnboarding.qml
Comment (brizental) [inline src/ui/views/ViewMobileOnboarding.qml:28]: On the Jira ticket, it says this page is the only one that has issues with bad contrast and I believed it. I did not manually check all pages.

Comment (lesleyjanenorton) [reply src/ui/views/ViewMobileOnboarding.qml:28]: So happy you picked this up! The Jira ticket didn't lie but what if we move this into `main.qml` so that it can be accessed from anywhere once we enable theming/dark mode? Or maybe we don't actually need the `VPNMobileStatusBarModifier` component and we just call `VPN.setStatusBarTextColor(...)` directly as necessary?

File: nebula/ui/components/VPNMobileStatusBarModifier.qml
Comment (bakulf) [inline nebula/ui/components/VPNMobileStatusBarModifier.qml:15]: If the enum is in the QML file, C++ needs to do what you have implemented: a quint8. And basically, we end up doing a mapping (int -> enum value). I wonder if it's cleaner to add an enum in C++ in `src/theme.h` and use the QT macro `Q_ENUM`. In this way, here we can do:

`property var statusBarTextColor: VPNTheme.StatusBarDarkTextColor` (or any better name).

We then propagate the VPNTheme.<NameOfTheEnum>.

File: nebula/ui/components/VPNMobileStatusBarModifier.qml
Comment (bakulf) [inline nebula/ui/components/VPNMobileStatusBarModifier.qml:19]: What about if we move this method into VPNTheme (theme.cpp/h)?

File: nebula/ui/components/VPNMobileStatusBarModifier.qml
Comment (bakulf) [inline nebula/ui/components/VPNMobileStatusBarModifier.qml:26]: Maybe we can set the status-bar color in the CTOR of theme.cpp. And reset it when the property changes.

Comment (brizental) [reply nebula/ui/components/VPNMobileStatusBarModifier.qml:26]: I implemented this component to overwrite whatever is the default status bar color. For the case described in the Jira ticket we want the text color to be light on onboarding pages and dark in other pages. That is why I chose to apply changes when the component is created and reset them when it's destroyed.

When there is a dark theme, it would be light in both and yes we would probably need to set that on the constructor of theme.cpp or main.qml. However, as I understand it, this is a different case.

File: src/ui/views/ViewMobileOnboarding.qml
Comment (brizental) [reply src/ui/views/ViewMobileOnboarding.qml:28]: So, IIUC this component should not be tied to the theme. I wrote it with the purpose to overwrite whatever is the theme default color for pages where a specific status bar text color is needed. That is why I only include it in this specific page, it is always going to have a dark background regardless of theme so it should reliably always have a light status bar text color.

The reason why I added the `VPNMobileStatusBarModifier` abstraction is because it includes the logic to apply the changes once the component is created and reset them once it is destroyed.

Let me know if the reasoning is incorrect :D

File: nebula/ui/components/VPNMobileStatusBarModifier.qml
Comment (bakulf) [reply nebula/ui/components/VPNMobileStatusBarModifier.qml:19]: This one. Which is called on completed.

File: nebula/ui/components/VPNMobileStatusBarModifier.qml
Comment (bakulf) [reply nebula/ui/components/VPNMobileStatusBarModifier.qml:26]: makes sense!

File: src/mozillavpn.h
Comment (bakulf) [inline src/mozillavpn.h:198]: Let's move these into the VPNTheme object (theme.cpp/h)

File: src/platforms/ios/iosutils.mm
Comment (brizental) [inline src/platforms/ios/iosutils.mm:124]: Not sure if this static_cast is too much of a hack. This is the most straight forward way I could find to have a setter for `preferredStatusBarStyle`. I am totally new to Objective-C so any input is appreciated :)

File: src/platforms/ios/iosutils.mm
Comment (brizental) [inline src/platforms/ios/iosutils.mm:15]: Reminder to self: Remove this unused include.

**Comment by brizental** [comment]
Hey folks, this is now ready for a final review. I did not add Android changes because this is an iOS only problem. On Android the status bar is always behing a dark background.

I spoke to @lesleyjanenorton about a bug I found on this implementation. Because I decided to apply changes when a given component is created and reset them when it is destroyed, the status bar color remains white if the user navigates from the onboarding page to the "Get Help" page. See video below.


https://user-images.githubusercontent.com/25176023/165277645-17941ec2-0f8a-444d-8a8e-54f46e9997f4.mov

Lesley mentioned that the work she is currently doing should fix this issue by making sure the Onboarding view is actually destroyed when users navigate to Get Help from it, so I believe the changes can be merged as is as we test it again once Lesley's changes are merged.

File: src/platforms/ios/iosutils.h
Comment (bakulf) [inline src/platforms/ios/iosutils.h:24]: Not need to be `const`

File: src/platforms/ios/iosutils.mm
Comment (bakulf) [inline src/platforms/ios/iosutils.mm:7]: alphabetic order here.

File: src/platforms/ios/iosutils.mm
Comment (bakulf) [reply src/platforms/ios/iosutils.mm:124]: No, it's actually correct. static_cast happens a compilation time.

File: src/theme.cpp
Comment (bakulf) [inline src/theme.cpp:153]: Nice! We often use a macro for this:
`Q_UNUSED(color);`

The code would be:
```
#ifdef MVPN_IOS
  ...
#else
  Q_UNUSED(color)
#endif
```

But I like your solution too!

**Review by bakulf — APPROVED** [review]

**Review by lesleyjanenorton — APPROVED** [review]

**Pull request merged**
